src/coreclr/vm/pendingload.h

```diff
@@ -17,248 +17,404 @@
+    }
 
 public:
-    PendingTypeLoadEntry(TypeKey typeKey, TypeHandle typeHnd)
-        : m_Crst(
-                 CrstPendingTypeLoadEntry,
-                 CrstFlags(CRST_HOST_BREAKABLE|CRST_UNSAFE_SAMELEVEL)
-                 ),
-        m_typeKey(typeKey)
-
+    struct Shard;
+    class Entry
     {
-        WRAPPER_NO_CONTRACT;
-
-        m_typeHandle = typeHnd;
-        m_dwWaitCount = 1;
-        m_hrResult = S_OK;
-        m_pException = NULL;
-#ifdef _DEBUG
-        if (LoggingOn(LF_CLASSLOADER, LL_INFO10000))
+        friend class PendingTypeLoadTable;
+    private:
+        Entry()
+            : m_Crst(
+                    CrstPendingTypeLoadEntry,
+                    CrstFlags(CRST_HOST_BREAKABLE|CRST_UNSAFE_SAMELEVEL)
+                    ),
+              m_typeKey(TypeKey::InvalidTypeKey()),
+              m_fIsPreallocated(true)
         {
-            SString name;
-            TypeString::AppendTypeKeyDebug(name, &m_typeKey);
-            LOG((LF_CLASSLOADER, LL_INFO10000, "PHASEDLOAD: Creating loading entry for type %s\n", name.GetUTF8()));
         }
-#endif
-
-        m_fLockAcquired = TRUE;
-
-        //---------------------------------------------------------------------------
-        // The PendingTypeLoadEntry() lock has a higher level than UnresolvedClassLock.
-        // But whenever we create one, we have to acquire it while holding the UnresolvedClassLock.
-        // This is safe since we're the ones that created the lock and are guaranteed to acquire
-        // it without blocking. But to prevent the crstlevel system from asserting, we
-        // must acquire using a special method.
-        //---------------------------------------------------------------------------
-        m_Crst.Enter(INDEBUG(Crst::CRST_NO_LEVEL_CHECK));
-    }
-
-    ~PendingTypeLoadEntry()
-    {
-        LIMITED_METHOD_CONTRACT;
 
-        if (m_fLockAcquired)
-            m_Crst.Leave();
-
-        if (m_pException && !m_pException->IsPreallocatedException()) {
-            delete m_pException;
+        Entry(Entry *pNext, DWORD hash, const TypeKey& typeKey, TypeHandle typeHnd)
+            : m_Crst(
+                    CrstPendingTypeLoadEntry,
+                    CrstFlags(CRST_HOST_BREAKABLE|CRST_UNSAFE_SAMELEVEL)
+                    ),
+              m_typeKey(typeKey),
+              m_fIsPreallocated(false)
+        {
+#ifdef PENDING_TYPE_LOAD_TABLE_STATS
+            PendingTypeLoadEntryDynamicAlloc();
+#endif // PENDING_TYPE_LOAD_TABLE_STATS
+            Init(pNext, hash, typeHnd);
         }
-    }
-
-#ifdef _DEBUG
-    BOOL HasLock()
-    {
-        LIMITED_METHOD_CONTRACT;
-        return m_Crst.OwnedByCurrentThread();
-    }
-#endif
 
-#ifndef DACCESS_COMPILE
-    VOID DECLSPEC_NORETURN ThrowException()
-    {
-        CONTRACTL
+        void SetTypeKey(const TypeKey& typeKey)
         {
-            THROWS;
-            GC_TRIGGERS;
-            INJECT_FAULT(COMPlusThrowOM(););
+            m_typeKey = typeKey;
```

Comment:
Is this copy desired? If so, it is more efficient to pass in the argument by value.